UI/MapWnd.cpp

```diff
@@ -4324,292 +4322,222 @@ std::pair<double, double> MapWnd::MovingFleetMapPositionOnLane(TemporaryPtr<cons
-        if (obj->SystemID() != INVALID_OBJECT_ID) {
-            ErrorLogger() << "a fleet that was supposed to be moving had a valid system in RefreshFleetButtons()";
-            continue;
-        }
+        // create new fleetbutton for this cluster of fleets
+        FleetButton* fb = new FleetButton(fleet_IDs, fleet_button_size);
 
-        // skip empty fleets
-        TemporaryPtr<const Fleet> fleet = boost::dynamic_pointer_cast<const Fleet>(obj);
-        if (fleet->Empty())
-            continue;
-
-        // store in map
-        moving_fleets[std::make_pair(obj->X(), obj->Y())][obj->Owner()].push_back(fleet);
-    }
-    moving_fleet_objects.clear();
+        // store per type of fleet button.
+        type_fleet_buttons[key].insert(fb);
 
+        // store for every fleet
+        for (std::vector<int>::const_iterator fleet_it = fleet_IDs.begin(); fleet_it != fleet_IDs.end(); ++fleet_it)
+            m_fleet_buttons[(*fleet_it)] = fb;
 
+        AttachChild(fb);
+        GG::Connect(fb->LeftClickedSignal,  boost::bind(&MapWnd::FleetButtonLeftClicked,    this, fb));
+        GG::Connect(fb->RightClickedSignal, boost::bind(&MapWnd::FleetButtonRightClicked,   this, fb));
+    }
+}
 
-    // clear old fleet buttons
+void MapWnd::DeleteFleetButtons() {
     m_fleet_buttons.clear();            // duplicates pointers in following containers
 
-    for (std::map<int, std::set<FleetButton*> >::iterator it = m_stationary_fleet_buttons.begin(); it != m_stationary_fleet_buttons.end(); ++it)
-        for (std::set<FleetButton*>::iterator set_it = it->second.begin(); set_it != it->second.end(); ++set_it)
+    for (boost::unordered_map<int, boost::unordered_set<FleetButton*> >::iterator it = m_stationary_fleet_buttons.begin();
+         it != m_stationary_fleet_buttons.end(); ++it)
+        for (boost::unordered_set<FleetButton*>::iterator set_it = it->second.begin(); set_it != it->second.end(); ++set_it)
             delete *set_it;
     m_stationary_fleet_buttons.clear();
 
-    for (std::map<int, std::set<FleetButton*> >::iterator it = m_departing_fleet_buttons.begin(); it != m_departing_fleet_buttons.end(); ++it)
-        for (std::set<FleetButton*>::iterator set_it = it->second.begin(); set_it != it->second.end(); ++set_it)
+    for (boost::unordered_map<int, boost::unordered_set<FleetButton*> >::iterator it = m_departing_fleet_buttons.begin();
+         it != m_departing_fleet_buttons.end(); ++it)
+        for (boost::unordered_set<FleetButton*>::iterator set_it = it->second.begin(); set_it != it->second.end(); ++set_it)
             delete *set_it;
     m_departing_fleet_buttons.clear();
 
-    for (std::set<FleetButton*>::iterator set_it = m_moving_fleet_buttons.begin(); set_it != m_moving_fleet_buttons.end(); ++set_it)
-        delete *set_it;
+    for (boost::unordered_map<std::pair<double, double>, boost::unordered_set<FleetButton*> >::iterator
+             it = m_moving_fleet_buttons.begin(); it != m_moving_fleet_buttons.end(); ++it)
```

Comment:
make indenting match start of boost:: on previous line
need { } after multi-line for